# データサイエンス特論 第13回 事前課題
青山学院大学大学院 理工学研究科 理工学専攻 知能情報コース 修士1年 35626302 森下剛

In [59]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

### ロジスティック回帰

#### 事前課題1の解答

In [60]:
# Iris データセットの読み込み
iris = load_iris()
X = iris.data
y = iris.target

# 全体の30%をテストデータに分割
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1, stratify=y)

#### 事前課題2の解答

**LogisticRegression の概要**

LogisticRegression は、ロジスティック回帰によるクラス分類を行う scikit-learn のクラスである。
- fit()：訓練データから係数を学習する。
- predict()：クラスを予測する。
- predict_proba()：確率値を得る。

単なる対数尤度の最大化ではなく、過学習を防ぐために正則化項を加えた目的関数を最適化しており、その挙動を以下のパラメータで制御できる。

**LogisticRegression のパラメータ**

- l1_ratio：Elastic-Net における L1 正則化の割合を 0 から 1 の実数で指定する。0 なら L2 正則化のみ、1 なら L1 正則化のみと同じになる。
- C：正則化の強さの逆数を指定する。C が小さいほど正則化が強く働き、大きいほど訓練データへの当てはまりが強くなる。
- solver：係数を求める際の最適化アルゴリズムを指定する。


#### 事前課題3の解答

In [61]:
# ロジスティック回帰モデルを学習
lr = LogisticRegression()
lr.fit(X_train, y_train)

# 予測と正解率の計算
y_pred = lr.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"accuracy = {accuracy}")

accuracy = 0.9777777777777777


#### 事前課題4の解答

In [62]:
# 予測結果の分割表を出力
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[15  0  0]
 [ 0 15  0]
 [ 0  1 14]]


In [63]:
# 構造の確認
print(iris.target_names)

['setosa' 'versicolor' 'virginica']


confusion_matrix の出力は、行が実際のクラス、列が予測クラスに対応している。

- setosa と versicolor は、各 15 件すべてが正しく分類された。
- virginica は 15 件中 14 件が正しく分類され、1 件が versicolor と誤分類された。

#### 事前課題5の解答

In [64]:
# 学習したモデルの係数と切片を確認
print(f"features: {iris.feature_names}")
print(f"coefficients:\n{lr.coef_}")
print(f"intercepts: {lr.intercept_}")

features: ['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)', 'petal width (cm)']
coefficients:
[[-0.4808825   0.7787016  -2.29149703 -0.92166338]
 [ 0.15174962 -0.21708714 -0.07817951 -0.69361031]
 [ 0.32913288 -0.56161446  2.36967654  1.61527369]]
intercepts: [  9.79209545   3.11181184 -12.90390729]


花びらの大きさに有意性が見られる。
petal length では setosa = -2.29 < versicolor = -0.07817951 < virginica = 2.36967654 となっている。つまり花びらの長さが小さいものは setosa で、大きいものは virginica で、中間が versicolor と解釈できる。
petal width でも同じ傾向が得られるため、花びらの大きさによってモデルが分類していると言える。